### PR TITLE
Feature: faster serialization for non nformat

### PR DESCRIPTION
**Pull request by liss-h**
Adds additional interfaces for the purpose of increasing serialization speed in non-nformat use-cases.

### CODE REVIEW

File: src/rdf4cpp/rdf/datatypes/registry/util/CharConvExt.hpp [hunk 1]
```diff
@@ -45,7 +45,7 @@ I from_chars(std::string_view s) {
  * @param value the value to be serialized
```
Comment (bigerl) [inline src/rdf4cpp/rdf/datatypes/registry/util/CharConvExt.hpp:46]: update doctstring

File: src/rdf4cpp/rdf/datatypes/registry/util/CharConvExt.hpp [hunk 2]
```diff
@@ -120,16 +121,16 @@ constexpr size_t log10ceil(T const value) noexcept {
...
```
Comment (bigerl) [inline src/rdf4cpp/rdf/datatypes/registry/util/CharConvExt.hpp:110]: update doctstring

File: src/rdf4cpp/rdf/datatypes/registry/util/CharConvExt.hpp [hunk 3]
```diff
@@ -192,7 +194,7 @@ std::string to_chars_canonical(F const value) noexcept {
...
```
Comment (bigerl) [inline src/rdf4cpp/rdf/datatypes/registry/util/CharConvExt.hpp:185]: update doctstring

File: src/rdf4cpp/rdf/writer/BufWriter.hpp [hunk 4]
```diff
@@ -157,36 +152,65 @@ struct BufWriterBase {
     [[nodiscard]] constexpr Cursor &cursor() noexcept { return cursor_; }
     [[nodiscard]] constexpr Buffer &buffer() noexcept { return buffer_; }
 
+    [[nodiscard]] constexpr Cursor const &cursor() const noexcept { return cursor_; }
+    [[nodiscard]] constexpr Buffer const &buffer() const noexcept { return buffer_; }
+
```
Comment (bigerl) [inline src/rdf4cpp/rdf/writer/BufWriter.hpp:186]: Is there any Writer that doesn't use BufWriterBase? Otherwise, we could also make it a requirement to use it as Base and add write_str as static method of BufWriterBase.

File: src/rdf4cpp/rdf/writer/BufWriter.hpp [hunk 5]
```diff
@@ -157,36 +152,65 @@ struct BufWriterBase {
...
     static void flush(void *buffer, Cursor *cursor, size_t additional_cap) noexcept {
         CRTP::flush_impl(*static_cast<typename CRTP::Buffer *>(buffer), *cursor, additional_cap);
     }
```
Comment (bigerl) [inline src/rdf4cpp/rdf/writer/BufWriter.hpp:189]: I would suggest to add a declaration of finalize() here as well and document it. Important aspect to be mentioned: It must be called only once (at least for some implementations).

File: src/rdf4cpp/rdf/query/Variable.cpp [hunk 6]
```diff
@@ -63,7 +63,7 @@ std::string_view Variable::name() const {
     return this->handle_.variable_backend().name;
 }
 
-bool Variable::serialize(void *const buffer, writer::Cursor *cursor, writer::FlushFunc const flush) const noexcept {
+bool Variable::serialize(void *const buffer, writer::Cursor *const cursor, writer::FlushFunc const flush) const noexcept {
```
Comment (bigerl) [inline src/rdf4cpp/rdf/query/Variable.cpp]: update interface in hpp as well

File: src/rdf4cpp/rdf/Quad.hpp [hunk 7]
```diff
@@ -56,23 +56,14 @@ class Quad : public query::QuadPattern {
...
-                *(it++) = item.to_node_storage(node_storage);
-        return qu;
-    }
+    [[nodiscard]] Quad to_node_storage(storage::node::NodeStorage &node_storage) const noexcept;
 
-    bool serialize_ntriples(void *buffer, writer::Cursor *cursor, writer::FlushFunc flush) const;
-    bool serialize_nquads(void *buffer, writer::Cursor *cursor, writer::FlushFunc flush) const;
-    bool serialize_turtle(writer::SerializationState& state, void *buffer, writer::Cursor *cursor, writer::FlushFunc flush) const;
-    bool serialize_trig(writer::SerializationState& state, void *buffer, writer::Cursor *cursor, writer::FlushFunc flush) const;
+    bool serialize_ntriples(void *buffer, writer::Cursor *cursor, writer::FlushFunc flush) const noexcept;
```
Comment (bigerl) [inline src/rdf4cpp/rdf/Quad.hpp]: *const missing in the signature compared to definition

File: src/rdf4cpp/rdf/writer/BufWriter.hpp
Comment (bigerl) [inline src/rdf4cpp/rdf/writer/BufWriter.hpp:1]: A genral interface thing: it is very common that a `BufWriter` is used to extract ` void *const buffer, Cursor *cursor, FlushFunc const flush`. That is enough because typically, not the full functionality is required for `write_str`. Similar interface are everywhere throughout the code. 

For the user, providing buffer,cursor,flush is cumbersome. I would suggest that we introduce a type that has that essential info as fields and is implicitly convertible from any BufWriter implementation. That way user's can just pass in their BufWriter. 

This can completely replace the overloads like
```
template<BufWriter W>
bool write_str(std::string_view str, W &w) {
    return write_str(str, &w.buffer(), &w.cursor(), &W::flush);
}
```

and allows us to only provide one simple interface like:

```
bool serialize(BufferWriterProxy w) { // ugly name I know; maybe you have a better suggestion :)
    ...
}
```

in user-facing types. it is not templated so implementation details can completely go into cpp files.

**Review by bigerl — CHANGES_REQUESTED** [review]
Thx for the PR. Things look good from what I can see. 
One thing I noticed, though. After going through the BufWriter architecture again for this review I found it felt a bit hard to follow. I have left some suggestions for improvements. Please start with the comments in BufWriter.

File: src/rdf4cpp/rdf/Quad.hpp (hunk 7 above)
Comment (liss-h) [reply src/rdf4cpp/rdf/Quad.hpp]: the outer most `const` does nothing on declarations

File: src/rdf4cpp/rdf/writer/BufWriter.hpp (hunk 4 above)
Comment (liss-h) [reply src/rdf4cpp/rdf/writer/BufWriter.hpp:186]: Every bufwriter that is not implemented here (in rdf4cpp) does not use it

File: src/rdf4cpp/rdf/writer/BufWriter.hpp
Comment (liss-h) [reply src/rdf4cpp/rdf/writer/BufWriter.hpp:1]: ~~Users are never really supposed to use the 3 param version, it simply exists for FFI purposes, adding the proxy defeats the purpose~~

Comment (liss-h) [reply src/rdf4cpp/rdf/writer/BufWriter.hpp:1]: I guess we could make it constructable from the three params as well

Comment (liss-h) [reply src/rdf4cpp/rdf/writer/BufWriter.hpp:1]: name suggestion: `BufWriterParts`

File: src/rdf4cpp/rdf/writer/BufWriter.hpp (hunk 5 above)
Comment (liss-h) [reply src/rdf4cpp/rdf/writer/BufWriter.hpp:189]: If we add a decl here we will get an IDE warning on every type inheriting from this (because the implementation in the derived will hide the previous one from the base)

File: src/rdf4cpp/rdf/Quad.hpp (hunk 7 above)
Comment (bigerl) [reply src/rdf4cpp/rdf/Quad.hpp]: it is compatible, yes. It's more that it makes me uneasy that the signatures are different.

File: src/rdf4cpp/rdf/writer/BufWriter.hpp (hunk 5 above)
Comment (bigerl) [reply src/rdf4cpp/rdf/writer/BufWriter.hpp:189]: How about CRTP for finalize as well?

File: src/rdf4cpp/rdf/Quad.hpp (hunk 7 above)
Comment (liss-h) [reply src/rdf4cpp/rdf/Quad.hpp]: Here are some arguments against using it in declarations:
1. It doesn't give the user any additional promises or any useful information at all
2. It doesn't help the compiler to optimize the code
3. It exposes an implementation detail that the users don't want to know and shouldn't know
4. It clutters the API surface with keywords that do nothing but make the function signature longer and harder to read
5. Declarations and definitions are already not symmetric in many other ways
   - `explicit` only appears in the declaration
   - `static` only appears in the declaration
   - `[[nodiscard]]` only appears in the declaration
   - `[[maybe_unused]]` only appears in defintions

Comment (bigerl) [reply src/rdf4cpp/rdf/Quad.hpp]: point 5. convinced me :)

File: src/rdf4cpp/rdf/writer/BufWriter.hpp (hunk 5 above)
Comment (liss-h) [reply src/rdf4cpp/rdf/writer/BufWriter.hpp:189]: I'm not sure what you are trying to achieve, but the `BufWriter` concept already tells you that you need to implement it and that you have to call it

Comment (bigerl) [reply src/rdf4cpp/rdf/writer/BufWriter.hpp:189]: OK. That's true. 
I am also not so sure I must admit. My main concern was that users are not seeing the documentation of finalize() in their IDE. But maybe a link in the class docstring of the concrete buffer Writer implementations to the BufWriter concept would do.

File: src/rdf4cpp/rdf/Node.hpp [hunk 8]
```diff
@@ -76,60 +76,22 @@ class Node {
...
-     * @param flush function to flush the contents of buffer and update cursor to point to the new free space
+     * @param parts writer parts
      * @return true if serialization was successful, false if a call to flush was not able to make room
      *
-     * For specific usage examples have a look at StringWriter and
-     * how the template Node::serialize<BufWriter> uses Serializers and tests/bench_SerDe.cpp#serialize.
+     * For specific usage examples have a look at tests/bench_SerDe.cpp#serialize.
      */
-    bool serialize(void *buffer, writer::Cursor *cursor, writer::FlushFunc flush) const noexcept;
+    bool serialize(writer::BufWriterParts parts) const noexcept;
```
Comment (bigerl) [inline src/rdf4cpp/rdf/Node.hpp]: I would call the parameter something like writer, buf_writer, writer_p, buf_writer_parts, writer_parts. (oder: most prefered -> least preferred).
It should focus more on what a user typically passes in then what it is converted to. After all the BufWriter implementations are mostly for construction/defining behaviour and are internally not used directly.

whatever you apply, please apply it everywhere

File: src/rdf4cpp/rdf/Node.hpp [hunk 9]
```diff
@@ -76,60 +76,22 @@ class Node {
...
-     * Takes as input an arbitrary, flushable/extendable buffer type (e.g. a std::string), a cursor
-     * into the buffer (that describes its base address and length of the free segment) and a function to flush the buffer (i.e. to either flush to io or extend the buffer).
-     * During operation will update the cursor to reflect the current start of the spare capacity and remaining length.
-     * When the remaining length reaches zero, flush is called using the provided buffer and cursor. Its task is to somehow make more buffer
-     * room and then adjust the cursor to point to the new spare room.
-     *
-     * @param buffer pointer to arbitrary buffer structure
-     * @param cursor cursor into buffer
-     * @param flush function to flush the contents of buffer and update cursor to point to the new free space
+     * @param parts writer parts
```
Comment (bigerl) [inline src/rdf4cpp/rdf/Node.hpp]:
```suggestion
     * @param writer Typically, you pass in an instance of a writer::BufWriter (Concept) instance. Any writer::BufWriter instance is implicitly convertible to writer::BufWriterParts.
```

File: src/rdf4cpp/rdf/datatypes/registry/util/CharConvExt.hpp [hunk 10]
```diff
@@ -56,21 +59,7 @@ bool to_chars_canonical(I const value, void *buffer, writer::Cursor *cursor, wri
...
- * Serializes an integer into its _simplified_ string representation which is e.g. used in
- * casting integers to strings.
- * @note for integers this is identical to the canonical representation
- *
- * @see https://www.w3.org/TR/xpath-functions/#casting-to-string
- *
- * @param value the value to be serialized
- */
-template<std::integral I>
-std::string to_chars_simplified(I const value) noexcept {
```
Comment (bigerl) [inline src/rdf4cpp/rdf/datatypes/registry/util/CharConvExt.hpp:71]: where did `to_chars_simplified` go?

**Review by bigerl — CHANGES_REQUESTED** [review]

File: src/rdf4cpp/rdf/datatypes/registry/util/CharConvExt.hpp (hunk 10 above)
Comment (liss-h) [reply src/rdf4cpp/rdf/datatypes/registry/util/CharConvExt.hpp:71]: Was unused

**Review by bigerl — APPROVED** [review]
